### PR TITLE
[TRTLLM-13490][feat] Support cross-attention with FlashInfer TRTLLM-Gen kernels on Blackwell

### PR DESCRIPTION
**Pull request by cascade812**
## Description

When TRTLLM_ENABLE_TRTLLM_GEN_ATTENTION=1, cross-attention currently falls back to THOP on both sm90 and sm100. This PR enables the `trtllm_gen_backend` path for cross-attention, which uses FlashInfer/TRTLLM-Gen kernels.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Summary by CodeRabbit

* **New Features**
  * Added cross-attention support with configurable KV-cache data types and enhanced cache management.
  * Extended attention backend with improved configuration validation and hardware compatibility checks across architectures.

* **Tests**
  * Added comprehensive cross-attention validation tests including hardware support verification and configuration constraint enforcement.

### CODE REVIEW

File: tests/integration/defs/llmapi/test_llm_api_pytorch_t5.py
Comment (yuxianq) [inline tests/integration/defs/llmapi/test_llm_api_pytorch_t5.py:1]: Please move this file to `tests/unittest/_torch/attention/`, `tests/unittest/_torch/attention_backend/` is not covered by CI, I have tried to move other tests in this directory to `tests/unittest/_torch/attention/` in https://github.com/NVIDIA/TensorRT-LLM/pull/15368

File: tensorrt_llm/_torch/attention_backend/fmha/flashinfer_trtllm_gen.py [hunk 1]
```diff
@@ -655,6 +664,11 @@ def is_supported(
         kv_cache_dtype, _ = self._get_kv_cache_dtype_and_total_blocks(meta, is_mla_enable)
         if kv_cache_dtype is None:
             kv_cache_dtype = torch_dtype_to_binding(q_dtype)
+        if meta.is_cross and kv_cache_dtype == DataType.NVFP4:
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/fmha/flashinfer_trtllm_gen.py]: Can we aggregate cross attention check to a single place?

File: tensorrt_llm/_torch/attention_backend/trtllm_gen.py [hunk 2]
```diff
@@ -912,6 +926,7 @@ def forward(
             params.seq_offset = seq_offset
             params.input_seq_length = max_context_q_len
             params.batch_size = num_seqs
+            params.cross_attention = meta.is_cross
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm_gen.py]: Don't need to set cross_attention separately for ctx and gen, we can set it when we create `FmhaParams` in https://github.com/NVIDIA/TensorRT-LLM/blob/381595702481e9884e0ecaf782bdb8df6b10fb41/tensorrt_llm/_torch/attention_backend/trtllm_gen.py#L891

File: tensorrt_llm/_torch/attention_backend/trtllm_gen.py [hunk 3]
```diff
@@ -1051,6 +1071,8 @@ def run_context(
             params.total_num_blocks,  # total_num_blocks
             params.kv_factor,  # kv_factor
             True,  # need_build_kv_cache_metadata
+            fwd.cross_kv if params.cross_attention and fwd.update_kv_cache else None,
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm_gen.py]: We only set `fwd.cross_kv` when `params.cross_attention and fwd.update_kv_cache`, so we can direclty use `fwd.cross_kv,  #cross_kv` here

File: tensorrt_llm/_torch/attention_backend/trtllm_gen.py [hunk 4]
```diff
@@ -1087,8 +1109,16 @@ def run_context(
             self._enable_pdl,  # enable_pdl
             kv_scale_pool,  # kv_scale_pool
             self.USE_SHARED_PAGED_KV_IDX,  # uses_shared_paged_kv_idx
+            (
+                False
+                if params.cross_attention
+                else AttentionMaskType(fwd.mask_type) == AttentionMaskType.causal
+            ),  # causal
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm_gen.py]: Can we set local variable `causal` instead of inlining it here?

File: tensorrt_llm/_torch/modules/cross_attention.py [hunk 5]
```diff
@@ -42,10 +42,11 @@ class CrossAttention(nn.Module):
...
     The cross-attention sub-layer honors ``ModelConfig.attn_backend``: when
-    set to ``"TRTLLM"`` it dispatches through the production C++ attention op
-    on every supported architecture. Cross-attention currently uses the THOP
-    attention path because the ``trtllm_gen`` backend API does not yet carry
-    encoder K/V tensors.
+    set to ``"TRTLLM"`` it dispatches through the production attention backend
+    on every supported architecture. If the internal ``trtllm_gen`` fast path
+    is enabled and supports the current shape, cross-attention writes encoder
+    K/V to the cross-KV pool and uses the trtllm-gen kernels; otherwise it
+    falls back to the standard THOP attention path.
```
Comment (yuxianq) [inline tensorrt_llm/_torch/modules/cross_attention.py]: This doc about internal `trtllm_gen` and THOP will soon be outdated, since we will refactor them to fmha libs (see https://github.com/NVIDIA/TensorRT-LLM/pull/15204), please remove the internal details.

File: tests/integration/defs/llmapi/test_llm_api_pytorch_bart.py [hunk 6]
```diff
@@ -63,6 +63,8 @@ def _test_case(
     num_return_sequences: int,
     exact_match: bool,
     feature_id: str,
+    marks=(),
```
Comment (yuxianq) [inline tests/integration/defs/llmapi/test_llm_api_pytorch_bart.py]: We never use non-default value for `marks`, can we remove it?

File: tests/integration/defs/llmapi/test_llm_api_pytorch_t5.py
Comment (cascade812) [reply tests/integration/defs/llmapi/test_llm_api_pytorch_t5.py:1]: tests/unittest/_torch/attention/ is good focused attention-module tests: small synthetic tensors, direct attention backend calls, no full model, no LLM, no model weights, no end-to-end generation.
So I think current folder is a better place because test_llm_api_pytorch_t5.py contains e2e LLM API tests.

Comment (yuxianq) [reply tests/integration/defs/llmapi/test_llm_api_pytorch_t5.py:1]: Sorry I mark the wrong file, I mean we can put `tests/unittest/_torch/attention_backend/test_trtllm_gen.py` to `tests/unittest/_torch/attention/`

Comment (cascade812) [reply tests/integration/defs/llmapi/test_llm_api_pytorch_t5.py:1]: Removed this file, as we have e2e tests which covers new changes in this PR.

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 7]
```diff
@@ -189,11 +189,11 @@ def max_context_length(self) -> int:
...
 
-        Cross-attention metadata is already expanded to one row per decoder
-        beam, and all beams read the same encoder K/V cache. Keep kernel beam
-        indirection disabled for that path.
+        Cross-attention K/V storage is request-scoped, but its metadata is
+        expanded to one row per decoder beam so kernels still need the active
+        decoder beam width.
         """
-        return 1 if self.is_cross else self.beam_width
+        return self.beam_width
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/trtllm.py]: If effective_beam_width == beam_width, why would we need effective_beam_width? Can we remove it and use beam_width instead?

File: tensorrt_llm/_torch/attention_backend/fmha/flashinfer_trtllm_gen.py [hunk 8]
```diff
@@ -862,6 +877,10 @@ def run_context(
         rope_params = attn.rope_params
         bmm1_scale_static = self._get_bmm1_scale(attn)
         attention_chunk_size = self._get_attention_chunk_size(attn)
+        if params.is_cross and fwd.update_kv_cache and fwd.cross_kv is None:
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/fmha/flashinfer_trtllm_gen.py]: Can we move this check to `_is_supported_with_reason`?

File: tensorrt_llm/_torch/attention_backend/fmha/flashinfer_trtllm_gen.py [hunk 9]
```diff
@@ -919,6 +938,8 @@ def run_context(
             params.total_num_blocks,  # total_num_blocks
             params.kv_factor,  # kv_factor
             True,  # need_build_kv_cache_metadata
+            fwd.cross_kv,
```
Comment (yuxianq) [inline tensorrt_llm/_torch/attention_backend/fmha/flashinfer_trtllm_gen.py]: Use comment to provide argument name

**Review by yuxianq — APPROVED** [review]

**Review by brb-nv — APPROVED** [review]
Changes to `tensorrt_llm/_torch/models/modeling_t5.py` look good to me.
